### PR TITLE
Added Support for trainpad

### PR DESCRIPTION
**Pull request by AnanyaKirti**
added support for trainpad

Add support for TrainPad

## Description

Added Support for trainpad

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
-[x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/trainpad/trainpad.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "quantum.h"
+
+#define LAYOUT(k000, k001, k002, k003) {{ k000,  k001,  k002,  k003 }}
```
Comment (elpekenin) [inline keyboards/trainpad/trainpad.h:21]: As per `config.h`, this should be defined on `info.json` and then this file deleted

**Review by elpekenin — CHANGES_REQUESTED** [review]

File: keyboards/trainpad/config.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#define DIODE_DIRECTION COL2ROW
+#define DIRECT_PINS {{C5, B3, B4, B5}}
```
[the remark this answers is not in the document]
Comment (AnanyaKirti) [reply keyboards/trainpad/config.h:21]: is it necessary?

[the remark this answers is not in the document]
Comment (drashna) [reply keyboards/trainpad/config.h:21]: Also, diode direction has no meaning with direct pins.

File: keyboards/trainpad/info.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    },
+	"processor": "atmega32u2",
+    "bootloader": "atmel-dfu",
+	"layouts": {
+        "LAYOUT": {
+            "layout": [
+                {"x": 0, "y": 0},
+                {"x": 1, "y": 0},
+                {"x": 2, "y": 0},
+                {"x": 3, "y": 0}
```
Comment (drashna) [inline keyboards/trainpad/info.json:19]:
```suggestion
    "matrix_pins": {
        "direct": [
            ["C5", "B3", "B4", "B5"],
        ]
    },
	"layouts": {
        "LAYOUT": {
            "layout": [
                {"x": 0, "y": 0, "matrix": [0, 0]},
                {"x": 1, "y": 0, "matrix": [0, 1]},
                {"x": 2, "y": 0, "matrix": [0, 2]},
                {"x": 3, "y": 0, "matrix": [0, 3]}
```

File: keyboards/trainpad/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+enum layer_names {
+    _BASE,
+    _FN
+};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+
+    [_BASE] = LAYOUT(KC_ENTER, KC_1, KC_2, KC_3),
+    [_FN] = LAYOUT(KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS),
```
Comment (drashna) [inline keyboards/trainpad/keymaps/default/keymap.c:27]:
```suggestion
```

File: keyboards/trainpad/trainpad.c
Comment (drashna) [inline keyboards/trainpad/trainpad.c:1]: File can be removed.

File: keyboards/trainpad/trainpad.h
Comment (drashna) [inline keyboards/trainpad/trainpad.h:1]: file can be removed.

File: keyboards/trainpad/config.h
Comment (drashna) [inline keyboards/trainpad/config.h:1]: file can be removed.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/trainpad/trainpad.h (hunk 1 above)
Comment (AnanyaKirti) [reply keyboards/trainpad/trainpad.h:21]: is it necessary to move it to info.json? can you suggest changes? I'll commit them

File: keyboards/trainpad/readme.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+* Hardware Supported: custom stacked FR4 case
+* Hardware Availability: Contact bob at IMKC Discord
+
+Make example for this keyboard (after setting up your build environment):
+
+    make TrainPad:default
+
+Flashing example for this keyboard:
+
+    make TrainPad:default:flash
```
[the remark this answers is not in the document]
Comment (AnanyaKirti) [reply keyboards/trainpad/readme.md:15]: I have a pcb for this, it's not hardwired, hence did not put it in hardwired

File: keyboards/trainpad/trainpad.c
Comment (AnanyaKirti) [reply keyboards/trainpad/trainpad.c:1]: done!

File: keyboards/trainpad/config.h
Comment (AnanyaKirti) [reply keyboards/trainpad/config.h:1]: done!

File: keyboards/trainpad/trainpad.h
Comment (AnanyaKirti) [reply keyboards/trainpad/trainpad.h:1]: done!

File: keyboards/trainpad/info.json [hunk 6]
```diff
@@ -0,0 +1,28 @@
+{
+    "keyboard_name": "TrainPad",
+	"manufacturer": "Ananya Kirti",
+	"url": "",
+	"maintainer": "Ananya Kirti",
```
Comment (noroadsleft) [inline keyboards/trainpad/info.json:5]: This is meant to be the maintainer's GitHub username.

```suggestion
	"maintainer": "AnanyaKirti",
```

File: keyboards/trainpad/info.json [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    "usb": {
+        "vid": "0x416B",
+        "pid": "0x0003",
+        "device_version": "0.0.1"
+    },
+	"processor": "atmega32u2",
```
Comment (noroadsleft) [inline keyboards/trainpad/info.json:11]: Moving these out of `rules.mk` to here.

```suggestion
    "features": {
        "bootmagic": true,
        "mousekey": true,
        "extrakey": true,
        "console": false,
        "command": false,
        "nkro": false
    },
    "processor": "atmega32u2",
```

File: keyboards/trainpad/keymaps/via/keymap.c [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT(
+    KC_ENT,  KC_1,  KC_2,  KC_3),
```
Comment (noroadsleft) [inline keyboards/trainpad/keymaps/via/keymap.c:20]: `via` keymaps actually should have four layers in most cases, even if the higher layers are fully transparent. VIA copies the keymap programmed in flash memory to EEPROM on first boot, and expects four layers by default. Without the last three layers, VIA will copy bad data into EEPROM, which leads to unexpected behaviour if the end user uses a VIA client to add keys to the higher layers.

```suggestion
    [0] = LAYOUT(
        KC_ENT,  KC_1,  KC_2,  KC_3
    ),
    [1] = LAYOUT(
        _______, _______, _______, _______
    ),
    [2] = LAYOUT(
        _______, _______, _______, _______
    ),
    [3] = LAYOUT(
        _______, _______, _______, _______
    ),
```

File: keyboards/trainpad/rules.mk [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+#
+BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+NKRO_ENABLE = no           # Enable N-Key Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+AUDIO_ENABLE = no           # Audio output
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
```
Comment (noroadsleft) [inline keyboards/trainpad/rules.mk:12]: Setting these through the `features` key in `info.json` means they're not needed here, but the file still must be present.

```suggestion
# This file is intentionally left blank
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]

File: keyboards/trainpad/keymaps/via/keymap.c (hunk 8 above)
Comment (noroadsleft) [reply keyboards/trainpad/keymaps/via/keymap.c:20]: Ahh, missed this one. Thanks @lesshonor.

File: keyboards/trainpad/readme.md [hunk 10]
```diff
@@ -0,0 +1,17 @@
+# trainpad
+
+Custom handwired one key keyboard. **See each individual board for pin information.**
```
[the remark this answers is not in the document]
Comment (AnanyaKirti) [reply keyboards/trainpad/readme.md]: corrected the hardwire part, it has a PCB

File: keyboards/trainpad/readme.md [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+* Keyboard Maintainer: [Ananya Kirti](https://github.com/AnanyaKirti)
+* Hardware Supported: custom stacked FR4 case
+* Hardware Availability: Contact bob at IMKC Discord
```
[the remark this answers is not in the document]
Comment (AnanyaKirti) [reply keyboards/trainpad/readme.md]: done!

File: keyboards/trainpad/trainpad.h (hunk 1 above)
Comment (AnanyaKirti) [reply keyboards/trainpad/trainpad.h]: committed the changes! Kindly let me know if they are fine!

File: keyboards/trainpad/readme.md [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+Make example for this keyboard (after setting up your build environment):
+
+    make trainpad:default
+
+Flashing example for this keyboard:
+
+    make trainpad:default:flash
+
+See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
```
Comment (noroadsleft) [inline keyboards/trainpad/readme.md:17]: Could you add instructions above this paragraph on how to enter bootloader mode for this keyboard?

Here's some sample text you can modify as needed for this purpose:

```
## Bootloader

Enter the bootloader in 3 ways:

* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
* **Physical reset button**: Briefly press the button on the back of the PCB - some may have pads you must short instead
* **Keycode in layout**: Press the key mapped to `QK_BOOT` if it is available
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

File: keyboards/trainpad/readme.md (hunk 12 above)
Comment (AnanyaKirti) [reply keyboards/trainpad/readme.md:17]: done!

**Comment by AnanyaKirti** [comment]
@noroadsleft any other changes needed?

File: keyboards/trainpad/readme.md [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+* Keyboard Maintainer: [Ananya Kirti](https://github.com/AnanyaKirti)
+* Hardware Supported: custom stacked FR4 case
+* Hardware Availability: Contact bob at [IMKC Discord](https://discord.gg/imkc)
+
+
+
+## Bootloader
+
+Enter the bootloader:
+* **Physical reset button**: Briefly press the taticle button on the back of the PCB.
```
Comment (noroadsleft) [inline keyboards/trainpad/readme.md:14]: You have Bootmagic enabled, so there's another possible reset method. Also fixes a typo.

```suggestion
* **Bootmagic reset**: Hold down the far left key while connecting the USB cable.
* **Physical reset button**: Briefly press the tactile button on the back of the PCB.
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]

**Review by noroadsleft — APPROVED** [review]

**Comment by noroadsleft** [comment]
Looks good, thanks!

**Comment by noroadsleft** [comment]
For future reference, we recommend against committing to your `master` branch as you've done here, because pull requests from modified `master` branches can make it more difficult to keep your QMK fork updated. It is highly recommended for QMK development – regardless of what is being done or where – to keep your master updated, but **NEVER** commit to it. Instead, do all your changes in a branch (branches are basically free in Git) and issue PRs from your branches when you're developing.

There are instructions on how to keep your fork updated here:

[**Best Practices: Your Fork's Master: Update Often, Commit Never**](https://docs.qmk.fm/#/newbs_git_using_your_master_branch)

[Fixing Your Branch](https://docs.qmk.fm/#/newbs_git_resynchronize_a_branch) will walk you through fixing up your `master` branch moving forward. If you need any help with this just ask.

Thanks for contributing!

**Comment by AnanyaKirti** [comment]
thank you for this! Sorry for all the trouble and your kind effort. It has been a while since I have done this. Thank you for being patient!

**Pull request merged**